analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaEcommerceHelper.java

```diff
@@ -0,0 +1,115 @@
+      if (SEG_VIEWED_PRODUCT_CATEGORY.equalsIgnoreCase(track.event())) {
+        processProductCategory(track);
+      } else if (SEG_VIEWED_PRODUCT.equalsIgnoreCase(track.event())) {
+        processViewedProductName(track);
+        processProductCategory(track);
+      } else if (SEG_ADDED_PRODUCT.equalsIgnoreCase(track.event())) {
+        processAddedProductName(track);
+        processAddedProductCategory(track);
+      } else if (SEG_COMPLETED_ORDER.equalsIgnoreCase(track.event())) {
+        processCompletedOrderDiscount(track);
+      }
+    }
+  }
+
+  private void processCompletedOrderDiscount(TrackPayload track) {
+    String discountString = track.properties().getString(SEG_DISCOUNT_KEY);
+    if (!TextUtils.isEmpty(discountString)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, discountString);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    } else {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, "0");
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductCategory(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_CATEGORY, track.properties().category());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processViewedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_VIEWED_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processProductCategory(TrackPayload track) {
+
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Set<String> kahunaCategoriesSet
+              = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+        protected boolean removeEldestEntry(Entry<String, Boolean> eldest) {
+          return size() > MAX_CATEGORIES_VIEWED_ENTRIES;
+        }
+      });
+      if (KahunaAnalytics.getUserAttributes().containsKey(CATEGORIES_VIEWED)) {
+        String serializedCategories = KahunaAnalytics.getUserAttributes().get(CATEGORIES_VIEWED);
+        kahunaCategoriesSet.addAll(Arrays.asList(serializedCategories.split(",")));
```

Comment:
So if there is a track event without a category, then CATEGORIES_VIEWED = None (from the line below). Then if there is another track event with a category (say sports), because this joins the categories, it would be "NONE, Sports" which seems weird.